models/case.js

```diff
@@ -1,11 +1,58 @@
+const createCaseAudit = async (CaseAudit, previousValue, newValue, transaction) => {
+  const caseAuditRecord = {
+    caseId: newValue.id,
+    twilioWorkerId: newValue.twilioWorkerId,
+    previousValue,
+    newValue,
+  };
+  await CaseAudit.create(caseAuditRecord, { transaction });
+};
+
+const getContactIds = async caseInstance => {
+  const contacts = await caseInstance.getContacts();
+  return contacts.map(contact => contact.dataValues.id);
+};
+
 module.exports = (sequelize, DataTypes) => {
   const Case = sequelize.define('Case', {
     status: DataTypes.STRING,
     helpline: DataTypes.STRING,
     info: DataTypes.JSONB,
+    twilioWorkerId: DataTypes.STRING,
+  });
+
+  Case.associate = models => {
+    Case.hasMany(models.Contact, { foreignKey: 'caseId' });
+    Case.hasMany(models.CaseAudit, { foreignKey: 'caseId' });
+  };
+
+  Case.afterCreate('auditCaseHook', async (caseInstance, options) => {
+    const { CaseAudit } = caseInstance.sequelize.models;
+    const contactIds = await getContactIds(caseInstance);
+    const previousValue = null;
+    const newValue = {
+      ...caseInstance.dataValues,
+      contacts: contactIds,
+    };
+
+    await createCaseAudit(CaseAudit, previousValue, newValue, options.transaction);
   });
 
-  Case.associate = models => Case.hasMany(models.Contact, { foreignKey: 'caseId' });
+  Case.afterUpdate('auditCaseHook', async (caseInstance, options) => {
+    const { CaseAudit } = caseInstance.sequelize.models;
+    const contactIds = await getContactIds(caseInstance);
+    const previousValue = {
+      ...caseInstance.dataValues,
+      ...caseInstance.previous(),
```

Comment:
What was the outcome on how this would handle new keys that were added?  Did that not end up being an issue?

Reply:
I've investigated and `previous()` would bring any  `newKey` as `null`, so it won't be an issue.